store/tikv/latch/latch.go

```diff
@@ -0,0 +1,202 @@
+	return &Latches{
+		slots:        slots,
+		waitingQueue: make(map[int][]uint64),
+	}
+}
+
+// GenLock generates Lock for the transaction with startTS and keys.
+func (latches *Latches) GenLock(startTS uint64, keys [][]byte) Lock {
+	slots := make([]int, 0, len(keys))
+	for _, key := range keys {
+		slots = append(slots, latches.hash(key))
+	}
+	sort.Ints(slots)
+	if len(slots) <= 1 {
+		return NewLock(startTS, slots)
+	}
+	dedup := slots[:1]
+	for i := 1; i < len(slots); i++ {
+		if slots[i] != slots[i-1] {
+			dedup = append(dedup, slots[i])
+		}
+	}
+	return NewLock(startTS, dedup)
+}
+
+// hash return hash int for current key.
+func (latches *Latches) hash(key []byte) int {
+	return int(murmur3.Sum32(key)) & (len(latches.slots) - 1)
+}
+
+// Acquire tries to acquire the lock for a transaction.
+// It returns with stale = true when the transaction is stale(
+// when the lock.startTS is smaller than any key's last commitTS).
+func (latches *Latches) Acquire(lock *Lock) (success, stale bool) {
+	for lock.acquiredCount < len(lock.requiredSlots) {
+		slotID := lock.requiredSlots[lock.acquiredCount]
+		success, stale = latches.acquireSlot(slotID, lock.startTS)
+		if success {
+			lock.acquiredCount++
+			lock.waiting = false
+			continue
+		}
+		if !stale {
+			lock.waiting = true
+		}
+		return
+	}
+	return
+}
+
+// Release releases all latches owned by the `lock` and returns the wakeup list.
+// Preconditions: the caller must ensure the transaction is at the front of the latches.
+func (latches *Latches) Release(lock *Lock, commitTS uint64) (wakeupList []uint64) {
+	wakeupCount := lock.acquiredCount
+	if lock.waiting {
+		wakeupCount++
+	}
+	wakeupList = make([]uint64, 0, wakeupCount)
+	for id := 0; id < wakeupCount; id++ {
+		slotID := lock.requiredSlots[id]
+
+		if hasNext, nextStartTS := latches.releaseSlot(slotID, lock.startTS, commitTS); hasNext {
+			wakeupList = append(wakeupList, nextStartTS)
+		}
+	}
+	return
+}
+
+func (latches *Latches) releaseSlot(slotID int, startTS, commitTS uint64) (hasNext bool, nextStartTS uint64) {
+	latch := &latches.slots[slotID]
+	latch.Lock()
+	defer latch.Unlock()
+	if startTS != latch.waitingQueueHead {
+		panic(fmt.Sprintf("invalid front ts %d, latch:%#v", startTS, latch))
+	}
+	if latch.maxCommitTS < commitTS {
+		latch.maxCommitTS = commitTS
+	}
+	if !latch.hasWaiting {
+		latch.free()
+		return
+	}
+
+	latches.Lock()
+	waiting := latches.waitingQueue[slotID]
+	hasNext = true
+	nextStartTS = waiting[0]
+	if len(waiting) == 1 {
+		delete(latches.waitingQueue, slotID)
+		latch.hasWaiting = false
+	} else {
+		latches.waitingQueue[slotID] = waiting[1:]
+	}
+	latches.Unlock()
+	latch.waitingQueueHead = nextStartTS
+	return
+}
+
+func (latches *Latches) acquireSlot(slotID int, startTS uint64) (success, stale bool) {
+	latch := &latches.slots[slotID]
+	latch.Lock()
+	defer latch.Unlock()
+	if stale = latch.maxCommitTS > startTS; stale {
+		return
+	}
+	// Empty latch
+	if !latch.occupied() {
```

Comment:
acquired := latch.aquire(start_ts)
if acquired {
  return
}
...

Reply:
So we check stale outside latch while do acquire in latch? I do not think is a good style.
if we do this, I  think we should also implement latch.isStale(startTS), latch.setHasWaiting(), and so on.